map/gps_track.cpp

```diff
@@ -12,193 +12,144 @@
-
-  LazyInitFile();
-
-  // Write point to the file.
-  // If file exception happens, then drop file.
-  if (m_file)
   {
-    try
-    {
-      for (auto const & point : points)
-      {
-        size_t evictedId;
-        m_file->Append(point, evictedId);
-      }
-    }
-    catch (RootException & e)
-    {
-      LOG(LINFO, ("GpsTrackFile.Append has caused exception:", e.Msg()));
-      m_file.reset();
-    }
+    lock_guard<mutex> lg(m_dataGuard);
+    m_points.emplace_back(point);
   }
+  ExecTask();
+}
 
-  if (!m_collection)
-    return;
-
-  // Add points
-  pair<size_t, size_t> evictedIds;
-  pair<size_t, size_t> const addedIds = m_collection->Add(points, evictedIds);
-  if (addedIds.first == GpsTrackCollection::kInvalidId)
-    return; // nothing was added
-
-  if (!m_callback)
-    return;
-
-  size_t const addedCount = addedIds.second - addedIds.first + 1;
-  ASSERT_GREATER_OR_EQUAL(m_collection->GetSize(), addedCount, ());
-
-  // Not all points from infos could be added to collection due to timestamp consequence restriction.
-  // Get added points from collection.
-  vector<pair<size_t, location::GpsTrackInfo>> toAdd;
-  toAdd.reserve(addedCount);
-  m_collection->ForEach([&toAdd](location::GpsTrackInfo const & point, size_t id)->bool
+void GpsTrack::AddPoints(vector<TItem> const & points)
+{
   {
-    toAdd.emplace_back(id, point);
-    return true;
-  }, m_collection->GetSize() - addedCount);
-  ASSERT_EQUAL(toAdd.size(), addedCount, ());
-
-  m_callback(move(toAdd), evictedIds);
+    lock_guard<mutex> lg(m_dataGuard);
+    m_points.insert(m_points.end(), points.begin(), points.end());
+  }
+  ExecTask();
 }
 
 void GpsTrack::Clear()
 {
-  lock_guard<mutex> lg(m_guard);
-
-  LazyInitFile();
-
-  if (m_file)
   {
-    try
-    {
-      m_file->Clear();
-    }
-    catch (RootException & e)
-    {
-      LOG(LINFO, ("GpsTrackFile.Clear has caused exception:", e.Msg()));
-      m_file.reset();
-    }
+    lock_guard<mutex> lg(m_dataGuard);
+    m_points.clear();
+    m_needClear = true;
   }
-
-  if (!m_collection)
-    return;
-
-  auto const evictedIds = m_collection->Clear();
-  if (evictedIds.first == GpsTrackCollection::kInvalidId)
-    return; // nothing was removed
-
-  if (!m_callback)
-    return;
-
-  m_callback(vector<pair<size_t, location::GpsTrackInfo>>(), evictedIds);
+  ExecTask();
 }
 
 void GpsTrack::SetDuration(hours duration)
 {
-  lock_guard<mutex> lg(m_guard);
-
-  m_duration = duration;
-
-  if (!m_collection)
-    return;
-
-  auto const evictedIds = m_collection->SetDuration(duration);
-  if (evictedIds.first == GpsTrackCollection::kInvalidId)
-    return; // nothing was removed
-
-  if (!m_callback)
-    return;
+  ASSERT_GREATER(duration.count(), 0, ());
 
-  m_callback(vector<pair<size_t, location::GpsTrackInfo>>(), evictedIds);
+  {
+    lock_guard<mutex> lg(m_dataGuard);
+    if (m_duration == duration)
+      return;
+    m_duration = duration;
+  }
+  if (HasCallback())
```

Comment:
Please, create HasCallbackImpl(), which returns true if callback was set, but checks it without lock. The reason is that you already has lock here, and there is no point to take it second time.